scripts/prepare/tsc.ts

```diff
@@ -0,0 +1,93 @@
+#!/usr/bin/env ../../node_modules/.bin/ts-node
+
+import { join } from 'path';
+import fs, { move } from 'fs-extra';
+import * as ts from 'typescript';
+import glob from 'glob';
+import { exec } from '../utils/exec';
+
+const hasFlag = (flags: string[], name: string) => !!flags.find((s) => s.startsWith(`--${name}`));
+
+const run = async ({ cwd, flags }: { cwd: string; flags: string[] }) => {
+  const {
+    bundler: { pre, post, tsConfig: tsconfigPath = 'tsconfig.json' },
+  } = await fs.readJson(join(cwd, 'package.json'));
+
+  if (pre) {
+    await exec(`node -r ${__dirname}/../node_modules/esbuild-register/register.js ${pre}`, { cwd });
+  }
+
+  const reset = hasFlag(flags, 'reset');
+  const watch = hasFlag(flags, 'watch');
+  // const optimized = hasFlag(flags, 'optimized');
+
+  if (reset) {
+    await fs.emptyDir(join(process.cwd(), 'dist'));
+  }
+
+  const content = ts.readJsonConfigFile(tsconfigPath, ts.sys.readFile);
+
+  const out = ts.parseJsonSourceFileConfigFileContent(
+    content,
+    {
+      useCaseSensitiveFileNames: true,
+      readDirectory: ts.sys.readDirectory,
+      fileExists: ts.sys.fileExists,
+      readFile: ts.sys.readFile,
+    },
+    process.cwd(),
+    {
+      noEmit: false,
+      outDir: join(process.cwd(), 'dist'),
+      target: ts.ScriptTarget.ES2020,
+      declaration: true,
+    }
+  );
+
+  const compilerOptions = out.options;
+
+  if (watch) {
+    const watcher = ts.createWatchProgram(
+      ts.createWatchCompilerHost(tsconfigPath, compilerOptions, ts.sys)
+    );
+    watcher.getProgram().emit();
+  } else {
+    ts.createProgram({
+      rootNames: out.fileNames,
+      options: { ...compilerOptions, module: ts.ModuleKind.ES2020, declaration: false },
+    }).emit();
+
+    const files = glob.sync(join(process.cwd(), 'dist', '*.js'));
+    await Promise.all(files.map((file) => move(file, file.replace('.js', '.mjs'), {})));
```

Comment:
I'm planning on improving this, the compiler host has the ability to override the writeFile method, which would allow me to write the file with a different extension than the one tsc wants to write... this would also be the way to get the watch-mode to work, most likely.